server/src/main/java/org/opensearch/action/search/TransportDeletePITAction.java

```diff
@@ -0,0 +1,172 @@
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+
+/**
+ * Transport action for deleting pit reader context - supports deleting list and all pit contexts
+ */
+public class TransportDeletePITAction extends HandledTransportAction<DeletePITRequest, DeletePITResponse> {
+    private SearchService searchService;
+    private final NamedWriteableRegistry namedWriteableRegistry;
+    private TransportSearchAction transportSearchAction;
+    private final ClusterService clusterService;
+    private final SearchTransportService searchTransportService;
+    private static final Logger logger = LogManager.getLogger(TransportDeletePITAction.class);
+
+    @Inject
+    public TransportDeletePITAction(
+        SearchService searchService,
+        TransportService transportService,
+        ActionFilters actionFilters,
+        NamedWriteableRegistry namedWriteableRegistry,
+        TransportSearchAction transportSearchAction,
+        ClusterService clusterService,
+        SearchTransportService searchTransportService
+    ) {
+        super(DeletePITAction.NAME, transportService, actionFilters, DeletePITRequest::new);
+        this.searchService = searchService;
+        this.namedWriteableRegistry = namedWriteableRegistry;
+        this.transportSearchAction = transportSearchAction;
+        this.clusterService = clusterService;
+        this.searchTransportService = searchTransportService;
+    }
+
+    @Override
+    protected void doExecute(Task task, DeletePITRequest request, ActionListener<DeletePITResponse> listener) {
+        List<SearchContextIdForNode> contexts = new ArrayList<>();
+        List<String> pitIds = request.getPitIds();
+        if (pitIds.size() == 1 && "_all".equals(pitIds.get(0))) {
+            deleteAllPits(listener);
+        } else {
+            for (String pitId : request.getPitIds()) {
+                SearchContextId contextId = SearchContextId.decode(namedWriteableRegistry, pitId);
+                contexts.addAll(contextId.shards().values());
+            }
+            deletePits(contexts, ActionListener.wrap(r -> {
+                if (r == contexts.size()) {
+                    listener.onResponse(new DeletePITResponse(true));
+                } else {
+                    logger.debug(
+                        () -> new ParameterizedMessage("Delete PITs failed. " + "Cleared {} contexts out of {}", r, contexts.size())
+                    );
+                    listener.onResponse(new DeletePITResponse(false));
+                }
+            }, e -> {
+                logger.debug("Delete PITs failed ", e);
+                listener.onResponse(new DeletePITResponse(false));
+            }));
+        }
+    }
+
+    /**
+     * Delete all active PIT reader contexts
+     */
+    void deleteAllPits(ActionListener<DeletePITResponse> listener) {
```

Comment:
can we move this to a separate handler class which would enable us to do better functional testing?